installer/downloader.js

```
@@ -1,10 +1,74 @@
-var path = require("path"),
-url = require("url"),
-upgradeLimiter = requireRoot("installer/upgrade-limiter.js"),
-{platform, network} = require("rise-common-electron"),
-moduleCommon = require("common-display-module"),
-config = requireRoot("installer/config.js"),
-version = requireRoot("version.json");
+const path = require("path");
+const url = require("url");
+const upgradeLimiter = requireRoot("installer/upgrade-limiter.js");
+const {platform, network} = require("rise-common-electron");
+const moduleCommon = require("common-display-module");
+const config = requireRoot("installer/config.js");
+const version = requireRoot("version.json");
+
+const MAX_DOWNLOAD_RETRIES = 4;
+
+let downloadContext = { count: 0 };
+
+function finishDownload(promiseResolver, value) {
+  promiseResolver(value);
+
+  downloadContext = { count: 0 };
+}
+
+function tryDownload() {
+  downloadContext.count += 1;
+
+  const currentPromise = downloadContext.promise =
+    network.downloadFile(downloadContext.module.url, downloadContext.savePath);
+
+  // only resolve or reject if the promise hasn't changed
+  currentPromise.then(outcome =>
+    downloadContext.promise === currentPromise &&
+    finishDownload(downloadContext.resolve, outcome)
+  )
+  .catch(error =>
+    downloadContext.promise === currentPromise &&
+    finishDownload(downloadContext.reject, error)
+  );
+}
+
+function downloadFile(remoteModule, savePath) {
+  if (downloadContext.count > 0 ) {
+    throw new Error("A download is already in progress");
```

@santiagonoguez Can you clarify why we should throw an error here and not just `return` instead?

---

or were you asking why not ?
    return Promise.reject(new Error("A download is already in progress"));

---

@santiagonoguez Yes wondering why halt halt execution by throwing an Error instead of just returning rejection.

---

@stulees changed it

old habits :D